riptide-httpclient/src/main/java/org/zalando/riptide/httpclient/EndOfStreamDetectingInputStream.java

```diff
@@ -0,0 +1,54 @@
+package org.zalando.riptide.httpclient;
+
+import org.zalando.fauxpas.ThrowingBiConsumer;
+
+import java.io.FilterInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+
+final class EndOfStreamDetectingInputStream extends FilterInputStream {
+
+    private boolean ended;
+    private final ThrowingBiConsumer<InputStream, Boolean, IOException> closer;
+
+    EndOfStreamDetectingInputStream(final InputStream in,
+            final ThrowingBiConsumer<InputStream, Boolean, IOException> closer) {
+        super(in);
+        this.closer = closer;
+    }
+
+    @Override
+    public int read() throws IOException {
+        final int read = super.read();
+
+        if (read == -1) {
+            ended = true;
+        }
+
+        return read;
+    }
+
+    @Override
+    public int read(byte[] b) throws IOException {
+        return onRead(super.read(b), b.length);
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        return onRead(super.read(b, off, len), len);
+    }
+
+    private int onRead(final int read, final int max) {
+        if (read < max) {
```

Comment:
Is this guaranteed to indicate the end of the stream? I would have guessed that the number of bytes read depends on the underlying network buffers. Or is apache http client blocking until the byte array can be filled fully?

Reply:
Good point. I was just assuming it's reliable. I'll change it to check for `-1`, that will definitely work.

Reply:
Fixed in https://github.com/zalando/riptide/pull/580